### PR TITLE
changed YieldCondition to accept a timeout and return its status.

### PR DESCRIPTION
**Pull request by omniraptorr**

### CODE REVIEW

File: Coroutine.ttslua [hunk 1]
```diff
@@ -1,16 +1,28 @@
...
+---@overload fun(condition: fun():boolean):boolean
+function Coroutine.yieldCondition(condition, timeout)
     local co = coroutine.running()
 
-    Wait.condition(function()
-        coroutine.resume(co)
-    end, condition)
+    local timeOutFunc
+    if type(timeout) == "number" then
+        timeOutFunc = function()
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: EmmyLua doesn't like this assignment, raising the error:

> Type mismatch. Required: 'void' Found: 'fun(): void' 

This occurs because `timeOutFunc` is not declared with a type or defined where declared. You'd need to add the type `nil | fun(): void` at the declaration site so EmmyLua knows this assignment is valid.

File: Coroutine.ttslua [hunk 2]
```diff
@@ -1,16 +1,28 @@
...
+---@param timeout number|nil @Timeout in seconds (optional).
+---@return boolean @Return whether the condition was met or not.
+---@overload fun(condition: fun():boolean):boolean
+function Coroutine.yieldCondition(condition, timeout)
     local co = coroutine.running()
 
-    Wait.condition(function()
-        coroutine.resume(co)
-    end, condition)
+    local timeOutFunc
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: I'm going to propose this be removed. However, if it were to stay, it's probably best to try match the TTS API docs (and tts-types) re: naming. So this would ideally be `timeoutFunc` instead of `timeOutFunc`.

File: Coroutine.ttslua [hunk 3]
```diff
@@ -1,16 +1,28 @@
...
+            coroutine.resume(co, false)
+        end
+    end
 
-    coroutine.yield()
+    local function resume()
+        coroutine.resume(co, true)
+    end
+
+    Wait.condition(resume, condition, timeout, timeOutFunc)
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: EmmyLua will not accept the latter two parameters.

`timeout` because:
> Type mismatch. Required: 'number' Found: 'nil|number' 

`timeoutFunc`/`timeOutFunc` because it presently has the type `void`, as above. However, even if `timeOutFunc` were given the type `nil | fun(): void`, you'd then get:

> Type mismatch. Required: 'fun(): void' Found: 'fun(): void|nil'

Basically, both errors come down to EmmyLua making this API strongly typed. In particular there's no overload that will ever allow a `nil` `timeout` because despite the API accepting that, it's not safe.

The reason it's not safe is that if `nil` were a valid value for `timeout`, someone could accidentally provide `nil` for `timeout` _and_ a non-nil `timeoutFunc` that'll never be called.

File: Coroutine.ttslua [hunk 4]
```diff
@@ -1,16 +1,28 @@
...
+---@return boolean @Return whether the condition was met or not.
+---@overload fun(condition: fun():boolean):boolean
+function Coroutine.yieldCondition(condition, timeout)
     local co = coroutine.running()
 
-    Wait.condition(function()
-        coroutine.resume(co)
-    end, condition)
+    local timeOutFunc
+    if type(timeout) == "number" then
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: There's no need to perform a type check and string comparison here for a few reason:

1. These APIs have EmmyLua types, so ideally if users are really concerned about type safety, they'll just use EmmyLua.

2. However, if a user does provide a string (or some other type) by accident (i.e. not using EmmyLua, or a stuffed up type cast etc.), this will silently cause the `timeoutFunc` to never be called, where as a user presumably _meant_ to provide a valid timeout.
    
    In general, it's best to "fail fast". That could mean implementing our own error reporting here when we encounter the wrong type - however that's a bit tedious. Instead if we jus let incorrect types flow through to MoonSharp, it will raise an error about it being the wrong type. Which is a good outcome, as developers will know they messed up.

As such, a simple truthy (not `nil`) check is what you'd want here i.e.

```lua
if timeout then
```

File: Coroutine.ttslua [hunk 5]
```diff
@@ -1,16 +1,28 @@
...
+    end
 
-    coroutine.yield()
+    local function resume()
+        coroutine.resume(co, true)
+    end
+
+    Wait.condition(resume, condition, timeout, timeOutFunc)
+
+    return coroutine.yield()
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: You need a EmmyLua typecast on this line, because `coroutine.yield()` can in _theory_ return _any_ value (or none at all); where as this method is documented as returning a `boolean` only.

File: Coroutine.ttslua [hunk 6]
```diff
@@ -1,16 +1,28 @@
 ---@class ge_tts__Coroutine
 local Coroutine = {}
 
---- Yields from the current coroutine, which will later be resumed once a condition is met.
+--- Yields from the current coroutine, to be resumed once a condition is met or an optional timeout limit is reached.
 ---@param condition fun(): boolean @Return true when the current coroutine should be resumed.
-function Coroutine.yieldCondition(condition)
+---@param timeout number|nil @Timeout in seconds (optional).
+---@return boolean @Return whether the condition was met or not.
+---@overload fun(condition: fun():boolean):boolean
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: I must admit, I've been inconsistent with this myself, however, I'd prefer spaces to follow `:`.

Also, if you could put `@overload`s at the top (above `@params` and `@return`, but below the description), that'd be appreciated, as it makes overloads consistently discoverable. Aside from consistency, it'd good to ensure overloads are prominent as they tend to be used more often than the main signature.

However, the return type of this overload should actually be `true`, not `boolean` i.e.

```lua
---@overload fun(condition fun(): boolean): true
```

The reason for this is that this function will _never_ return `false` if the `timeout` parameter is omitted. This is important information, as it lets consumers know they can safely ignore the return value _unless_ they're providing a `timeout`.

File: Coroutine.ttslua [hunk 7]
```diff
@@ -1,16 +1,28 @@
 ---@class ge_tts__Coroutine
 local Coroutine = {}
 
---- Yields from the current coroutine, which will later be resumed once a condition is met.
+--- Yields from the current coroutine, to be resumed once a condition is met or an optional timeout limit is reached.
 ---@param condition fun(): boolean @Return true when the current coroutine should be resumed.
-function Coroutine.yieldCondition(condition)
+---@param timeout number|nil @Timeout in seconds (optional).
+---@return boolean @Return whether the condition was met or not.
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: Seems as this is documenting the return value, there's no need to include "Return" in the description.

Additionally, the language "Whether or not" does not explain the "not" case. I'd suggest:

> true if the condition was met, or false if the (optional) timeout was reached.

File: Coroutine.ttslua [hunk 8]
```diff
@@ -1,16 +1,28 @@
 ---@class ge_tts__Coroutine
 local Coroutine = {}
 
---- Yields from the current coroutine, which will later be resumed once a condition is met.
+--- Yields from the current coroutine, to be resumed once a condition is met or an optional timeout limit is reached.
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: Not a fan of replacing:
> which will later

with:
> to

Just doesn't seem as specific about what's happening. Also a "timeout" is inherently a limit, you'd either say "time limit" or "timeout", however seems as there is a `timeout` parameter, just saying "timeout" is best.

File: Coroutine.ttslua [hunk 9]
```diff
@@ -1,16 +1,28 @@
...
-    end, condition)
+    local timeOutFunc
+    if type(timeout) == "number" then
+        timeOutFunc = function()
+            coroutine.resume(co, false)
+        end
+    end
 
-    coroutine.yield()
+    local function resume()
```
Comment (Benjamin-Dobell) [inline Coroutine.ttslua]: I tend to prefer avoiding defining variables if they're just used once. The exceptions are:

* Named constants e.g. Instead of just parsing seemingly arbitrary hard-coded numbers to functions, it's a good idea to have a variable that explains what the variable represents e.g. `MAX_HAND_SIZE` instead of just using `10`.
* A dynamic expression that is either long/complex, or is a value you're likely to want to see in a debugger e.g.
    ```lua
    local somethingDynamic = (aValue > someOtherValue or aNumber == favoriteNumber) and callSomething() or callSomethingElse()
    ```

In this case, it's easy just to pass this function straight to `Wait.condition`. There is an argument to be made for using names when a function take several similar looking arguments, however sane IDEs solve that problem for you:

Screen Shot 2020-06-10 at 3 32 13 am

Note the "chips" stating "func" and "conditionFunc", letting you know which is which.

**Review by Benjamin-Dobell — CHANGES_REQUESTED** [review]
TLDR:

```lua
--- Yields from the current coroutine, which will be resumed once a condition is met or an optional timeout is reached.
---@overload fun(condition: fun(): boolean): true
---@param condition fun(): boolean @Return true when the current coroutine should be resumed.
---@param timeout number|nil @Timeout in seconds (optional).
---@return boolean @true if the condition was met, or false if the (optional) timeout was reached.
function Coroutine.yieldCondition(condition, timeout)
    local co = coroutine.running()

    local conditionMet = function()
        coroutine.resume(co, true)
    end

    if timeout then
        Wait.condition(conditionMet, condition, --[[---@not nil]] timeout, function()
            coroutine.resume(co, false)
        end)
    else
        Wait.condition(conditionMet, condition)
    end

    return --[[---@type boolean]] coroutine.yield()
end
```

File: Coroutine.ttslua (hunk 5 above)
Comment (omniraptorr) [reply Coroutine.ttslua]: i don't understand this one, is this because something could preempt the scheduled resume (since we already scheduled a boolean to be passed into the resume)

Comment (Benjamin-Dobell) [reply Coroutine.ttslua]: Oh, nah. Nothing that fancy. `coroutine.yield` just has the EmmyLua doc type:

```lua
---@vararg any
---@return any...
function coroutine.yield(...) end
```

The return type is `any...` so, which could be _no_ return value at all (e.g. `nil`).

So when we call this function EmmyLua is basically going:

> Hmm, we've said this function returns a `boolean`, but `coroutine.yield()` may actually return `nil`. Nope, they're not compatible types.

_Technically_ EmmyLua is actually correct here as well. Because at some other stage, another piece of code _could_ have grabbed a hold of this coroutine (Lua `thread`) and call `coroutine.resume(ourCoroutine)` i.e. We _may_ actually get `nil` here, or well, any value at all!

However, in this particular case, I think it's fair to assume users won't do bonkers stuff like that, and thus assume that it's only us calling `coroutine.resume`. In which case, we _know_ we're resuming with either `true` or `false`, so we can "safely" (with the aforementioned caveats) cast to `boolean`.

**Review by Benjamin-Dobell — APPROVED** [review]
